src/view.c

```diff
@@ -18,43 +18,53 @@
 #define LAB_FALLBACK_WIDTH  640
 #define LAB_FALLBACK_HEIGHT 480
 
-void
-view_array_append(struct server *server, struct wl_array *views,
-		enum lab_view_criteria criteria)
+struct view *
+view_next(struct view *view, enum lab_view_criteria criteria)
 {
-	struct view *view;
-	wl_list_for_each(view, &server->views, link)
-	{
-		if (!isfocusable(view)) {
-			continue;
-		}
+	assert(view);
+	struct server *server = view->server;
 
-		if (criteria & LAB_VIEW_CRITERIA_CURRENT_WORKSPACE) {
-			/*
-			 * Always-on-top views are always on the current
-			 * desktop and are special in that they live in a
-			 * different tree.
-			 */
-			if (view_is_always_on_top(view)) {
-				goto next;
-			}
-			if (view->scene_tree->node.parent != server->workspace_current->tree) {
-				continue;
-			}
+	view = wl_container_of(view->link.next, view, link);
+	if (&view->link == &server->views) {
+		return view;
+	}
+	if (!isfocusable(view)) {
+		return view_next(view, criteria);
```

Comment:
Being a paranoid here, but given enough !isfocusable() views a malicious user can overflow the stack. Can we do something about it?

Reply:
Although this seems to reference and outdated state, the same comment still applies to the current state. Maybe we could just throw it into a `for(;;) {` loop, similar to `view_first()`?

Reply:
Good feedback. Loop is much better than recursion here. Have implemented.